src/Renci.SshNet/Security/Cryptography/Ciphers/AesGcmCipher.cs

```diff
@@ -0,0 +1,171 @@
+#else
+            _aesGcm = new AesGcm(key);
+#endif
+        }
+
+        /// <summary>
+        /// Encrypts the specified input.
+        /// </summary>
+        /// <param name="input">
+        /// The input data with below format:
+        ///   <code>
+        ///   [outbound sequence field][packet length field][padding length field sz][payload][random paddings]
+        ///   [----4 bytes----(offset)][------4 bytes------][----------------Plain Text---------------(length)]
+        ///   </code>
+        /// </param>
+        /// <param name="offset">The zero-based offset in <paramref name="input"/> at which to begin encrypting.</param>
+        /// <param name="length">The number of bytes to encrypt from <paramref name="input"/>.</param>
+        /// <returns>
+        /// The encrypted data with below format:
+        ///   <code>
+        ///   [packet length field][padding length field sz][payload][random paddings][Authenticated TAG]
+        ///   [------4 bytes------][------------------Cipher Text--------------------][-------TAG-------]
+        ///   </code>
+        /// </returns>
+        public override byte[] Encrypt(byte[] input, int offset, int length)
+        {
+            var packetLengthField = new ReadOnlySpan<byte>(input, offset, 4);
+            var plainText = new ReadOnlySpan<byte>(input, offset + 4, length - 4);
+
+            var output = new byte[length + TagSize];
+            packetLengthField.CopyTo(output);
+            var cipherText = new Span<byte>(output, 4, length - 4);
+            var tag = new Span<byte>(output, length, TagSize);
+
+            _aesGcm.Encrypt(nonce: _iv, plainText, cipherText, tag, associatedData: packetLengthField);
+
+            IncrementCounter();
+
+            return output;
+        }
+
+        /// <summary>
+        /// Decrypts the specified input.
+        /// </summary>
+        /// <param name="input">
+        /// The input data with below format:
+        ///   <code>
+        ///   [inbound sequence field][packet length field][padding length field sz][payload][random paddings][Authenticated TAG]
+        ///   [--------4 bytes-------][--4 bytes--(offset)][--------------Cipher Text----------------(length)][-------TAG-------]
+        ///   </code>
+        /// </param>
+        /// <param name="offset">The zero-based offset in <paramref name="input"/> at which to begin decrypting and authenticating.</param>
+        /// <param name="length">The number of bytes to decrypt and authenticate from <paramref name="input"/>.</param>
+        /// <returns>
+        /// The decrypted data with below format:
+        /// <code>
+        ///   [padding length field sz][payload][random paddings]
+        ///   [--------------------Plain Text-------------------]
+        /// </code>
+        /// </returns>
+        public override byte[] Decrypt(byte[] input, int offset, int length)
+        {
+            var packetLengthField = new ReadOnlySpan<byte>(input, 4, 4);
```

Comment:
Was it necessary to revert back to this way ? 🙁 It looked OK before

Reply:
I don't have a good understanding of chacha20-poly1305, but it sounds like it might be worth having a more general abstraction for building & deserializing packets, rather than trying to force these different schemes through the existing `Cipher` API.

For now, can we either put this bit back so that packet_length starts at offset, or just throw if `offset != 8` to make the behaviour more explicit. And then make AesGcmCipher internal so that it's harder to use directly? Then we can change it later

Reply:
That's a good idea.

Reply:
Actually, it might be good to set all ciphers to internal. They are SSH specific and are not intended to be consumed externally.
However, it's hard to change public APIs to internal as it is breaking change. We have no idea how they are being used.

Reply:
Using `Debug.Assert(offset == 8, "The offset must be 8");`  should be sufficient.

Reply:
Yes, let's leave the others. Aes is used externally, e.g. https://github.com/darinkes/SshNet.Keygen/blob/b230c43c53ceec442c3bb563957cd53bd03cac3e/SshNet.Keygen/SshKeyEncryption/SshKeyEncryptionAes256.cs#L74

Most of the rest I would like to delete